File: android/src/org/tmurakam/cashflow/models/Transaction.java
// -*-  Mode:java; c-basic-offset:4; tab-width:4; indent-tabs-mode:t -*-

package org.tmurakam.cashflow.models;

import java.util.*;

import org.tmurakam.cashflow.*;
import org.tmurakam.cashflow.ormapper.*;

public  class Transaction extends TransactionBase implements Cloneable {
	public final static int OUTGO = 0;
	public final static int INCOME = 1;
	public final static int ADJ = 2;
	public final static int TRANSFER = 3;

	public boolean hasBalance;
	public double balance;

	public final static Transaction instance = new Transaction();
		
	public Transaction() {
		this(0, "", 0.0);

		// 現在時刻をセット
		Calendar cal = new GregorianCalendar();
		AppConfig cfg = AppConfig.instance();
		if (cfg.dateTimeMode == AppConfig.DateTimeModeDateOnly) {
			// 時刻を 0:00:00 に設定
			cal.setTimeInMillis(this.date);
			cal.set(Calendar.HOUR_OF_DAY, 0);
			cal.set(Calendar.MINUTE, 0);
			cal.set(Calendar.SECOND, 0);
			cal.set(Calendar.MILLISECOND, 0);
		}
		this.date = cal.getTimeInMillis();
	}

	public Transaction(long date, String description, double value) {
		this.asset = -1;
		this.dst_asset = -1;
		this.date = date;
		this.description = description;
		this.memo = "";
		this.value = value;
		this.type = 0;
		this.category = -1;
		this.pid = 0; // init
		this.hasBalance = false;
	}

	// deep copy
	@Override
		public Object clone() {
		try {
			return super.clone();
		}
		catch (CloneNotSupportedException e) {
			throw new InternalError(e.toString());
		}
	}

	public void insert() {
		super.insert();
		DescLRUManager.addDescLRU(description, category);
	}

	public void update() {
		super.update();
		DescLRUManager.addDescLRU(description, category);
	}

	public void updateWithoutUpdateLRU() {
		super.update();
	}
}

File: Classes/models/Transaction.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2011, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

#import "Transaction.h"
#import "Database.h"
#import "Config.h"
#import "DescLRUManager.h"

@implementation Transaction

@synthesize hasBalance = mHasBalance;
@synthesize balance = mBalance;

/*
 */
+ (BOOL)migrate
{
    BOOL ret = [super migrate];
    return ret;
}

- (id)init
{
    [super init];

    self.asset = -1;
    self.dstAsset = -1;
    
    // 現在時刻で作成
    NSDate *dt = [[[NSDate alloc] init] autorelease];
    
    if ([Config instance].dateTimeMode == DateTimeModeDateOnly) {
        // 時刻を 0:00:00 に設定
        NSCalendar *greg = [[[NSCalendar alloc] initWithCalendarIdentifier:NSGregorianCalendar] autorelease];
        NSDateComponents *dc = [greg components:(NSYearCalendarUnit | NSMonthCalendarUnit | NSDayCalendarUnit) fromDate:dt];
        dt = [greg dateFromComponents:dc];
    }
    
    self.date = dt;
    self.description = @"";
    self.memo = @"";
    self.value = 0.0;
    self.type = 0;
    self.category = -1;
    self.hasBalance = NO;
    return self;
}

- (void)dealloc
{
    [super dealloc];
}

- (id)initWithDate: (NSDate*)dt description:(NSString*)desc value:(double)v
{
    [super init];

    self.asset = -1;
    self.dstAsset = -1;
    self.date = dt;
    self.description = desc;
    self.memo = @"";
    self.value = v;
    self.type = 0;
    self.category = -1;
    self.pid = 0; // init
    self.hasBalance = NO;
    return self;
}

- (id)copyWithZone:(NSZone*)zone
{
    Transaction *n = [[Transaction alloc] init];
    n.pid = self.pid;
    n.asset = self.asset;
    n.dstAsset = self.dstAsset;
    n.date = self.date;
    n.description = self.description;
    n.memo = self.memo;
    n.value = self.value;
    n.type = self.type;
    n.category = self.category;
    n.hasBalance = self.hasBalance;
    n.balance = self.balance;
    return n;
}

- (void)_insert
{
    [super _insert];
    [DescLRUManager addDescLRU:self.description category:self.category];
}

- (void)_update
{
    [super _update];
    [DescLRUManager addDescLRU:self.description category:self.category];
}

- (void)updateWithoutUpdateLRU
{
    [super _update];
}

@end
